packages/local_auth/local_auth_darwin/darwin/local_auth_darwin/Sources/local_auth_darwin/LocalAuthPlugin.swift

```diff
@@ -0,0 +1,254 @@
+        DispatchQueue.main.async {
+          self.handleAuthReply(
+            success: success, error: error as NSError?, options: options, strings: strings,
+            completion: completion)
+        }
+      }
+    } else {
+      let error = authError ?? NSError(domain: "", code: 0, userInfo: nil)
+      handleError(authError: error, options: options, strings: strings, completion: completion)
+    }
+  }
+
+  func handleAuthReply(
+    success: Bool,
+    error: NSError?,
+    options: AuthOptions,
+    strings: AuthStrings,
+    completion: @escaping (Result<AuthResultDetails, Error>) -> Void
+  ) {
+    assert(Thread.isMainThread, "Response handling must be done on the main thread.")
+
+    if success {
+      handleSucceeded(succeeded: true, completion: completion)
+    } else {
+      guard let error = error else {
+        handleError(
+          authError: NSError(domain: error!.domain, code: error!.code, userInfo: error?.userInfo),
+          options: options,
+          strings: strings,
+          completion: completion
+        )
+        return
+      }
+
+      switch error.code {
+      case LAError.biometryNotAvailable.rawValue,
+        LAError.biometryNotEnrolled.rawValue,
+        LAError.biometryLockout.rawValue,
+        LAError.userFallback.rawValue,
+        LAError.passcodeNotSet.rawValue,
+        LAError.authenticationFailed.rawValue:
+        handleError(authError: error, options: options, strings: strings, completion: completion)
+        return
+
+      case LAError.systemCancel.rawValue:
+        if options.sticky {
+          lastCallState = StickyAuthState(
+            options: options, strings: strings, resultHandler: completion)
+        } else {
+          handleSucceeded(succeeded: false, completion: completion)
+        }
+        return
+
+      default:
+        handleError(authError: error, options: options, strings: strings, completion: completion)
+      }
+    }
+  }
+
+  func handleSucceeded(
+    succeeded: Bool,
+    completion: @escaping (Result<AuthResultDetails, Error>) -> Void
+  ) {
+    let result: AuthResult = succeeded ? .success : .failure
+    let resultDetails = AuthResultDetails(result: result, errorMessage: nil, errorDetails: nil)
+    completion(.success(resultDetails))
+  }
+
+  func handleError(
+    authError: NSError,
+    options: AuthOptions,
+    strings: AuthStrings,
+    completion: @escaping (Result<AuthResultDetails, Error>) -> Void
+  ) {
+    var result: AuthResult = .errorNotAvailable
+
+    switch authError.code {
+    case LAError.passcodeNotSet.rawValue, LAError.biometryNotEnrolled.rawValue:
+      if options.useErrorDialogs {
+        alertController.showAlert(
+          message: strings.goToSettingsDescription ?? "",
+          dismissTitle: strings.cancelButton,
+          openSettingsTitle: strings.goToSettingsButton
+        ) { success in
+          let result: Result<AuthResultDetails, Error> =
```

Comment:
The protocol for showing dialogs was changed to handle authentication results using Result<AuthResultDetails, Error> instead of a simple Bool completion. And also because I changed the encapsulation of the logic (Default AlertController for iOS and macOS), instead of using a Factory.

Reply:
> And also because I changed the encapsulation of the logic

You've moved significant logic into a class that is replaced in tests, which means all of that logic is now untested, reducing test coverage relative to the previous approach. There is a reason we make the injectable test replacements absolutely minimal wrappers, rather than high-level abstractions.

Reply:
I'm just a bit confused about how to work with the alerts while maintaining what we had before. I’ll push some changes later today or tomorrow with these updates.